shell/browser/api/electron_api_web_contents.cc

```diff
@@ -3450,37 +3451,118 @@ v8::Local<v8::Promise> WebContents::TakeHeapSnapshot(
-  granted_devices_[render_frame_host->GetFrameTreeNodeId()][permissionType]
+  granted_devices_[render_frame_host->GetFrameTreeNodeId()][permission_type]
                   [origin]
                       .push_back(
                           std::make_unique<base::Value>(device->Clone()));
 }
 
-std::vector<base::Value> WebContents::GetGrantedDevices(
+void WebContents::RevokeDevicePermission(
     const url::Origin& origin,
-    content::PermissionType permissionType,
+    const base::Value* device,
+    content::PermissionType permission_type,
+    content::RenderFrameHost* render_frame_host) {
+  const auto& devices_for_frame_host_it =
+      granted_devices_.find(render_frame_host->GetFrameTreeNodeId());
+  if (devices_for_frame_host_it == granted_devices_.end())
+    return;
+
+  const auto& current_devices_it =
+      devices_for_frame_host_it->second.find(permission_type);
+  if (current_devices_it == devices_for_frame_host_it->second.end())
+    return;
+
+  const auto& origin_devices_it = current_devices_it->second.find(origin);
+  if (origin_devices_it == current_devices_it->second.end())
+    return;
+
+  for (auto it = origin_devices_it->second.begin();
+       it != origin_devices_it->second.end();) {
+    if (DoesDeviceMatch(device, it->get(), permission_type)) {
+      it = origin_devices_it->second.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+bool WebContents::DoesDeviceMatch(const base::Value* device,
+                                  const base::Value* device_to_compare,
+                                  content::PermissionType permission_type) {
+  if (permission_type ==
+      static_cast<content::PermissionType>(
+          WebContentsPermissionHelper::PermissionType::HID)) {
+    if (device->FindIntKey(kHidVendorIdKey) !=
+            device_to_compare->FindIntKey(kHidVendorIdKey) ||
+        device->FindIntKey(kHidProductIdKey) !=
+            device_to_compare->FindIntKey(kHidProductIdKey)) {
+      return false;
+    }
+
+    const auto* serial_number =
+        device_to_compare->FindStringKey(kHidSerialNumberKey);
+    const auto* device_serial_number =
+        device->FindStringKey(kHidSerialNumberKey);
+
+    if (serial_number && device_serial_number &&
+        *device_serial_number == *serial_number)
+      return true;
+  } else if (permission_type ==
+             static_cast<content::PermissionType>(
+                 WebContentsPermissionHelper::PermissionType::SERIAL)) {
+#if BUILDFLAG(IS_WIN)
+    if (device->FindStringKey(kDeviceInstanceIdKey) ==
+        device_to_compare->FindStringKey(kDeviceInstanceIdKey))
+      return true;
+#else
+    if (device->FindIntKey(kVendorIdKey) !=
+            device_to_compare->FindIntKey(kVendorIdKey) ||
+        device->FindIntKey(kProductIdKey) !=
+            device_to_compare->FindIntKey(kProductIdKey) ||
+        *device->FindStringKey(kSerialNumberKey) !=
+            *device_to_compare->FindStringKey(kSerialNumberKey)) {
```

Comment:
This will crash with a nullptr dereference if either FindStringKey() call fails. Are these guaranteed to return  non-nullptr?

Reply:
Yes.  The only time this code path is exercised is when [HidChooserContext::CanStorePersistentEntry](https://github.com/electron/electron/blob/main/shell/browser/hid/hid_chooser_context.cc#L58) is true and that checks to make sure a serial number is present.